Added a hook for simply adding locomotion

taeuscherpferd opened the pull request:
# Changes
- Added a hook that allows developers to quickly get up and running with simple locomotion
- Updated the documentation on how to use the new hook

bbohlender commented:
Hey @taeuscherpferd 
sorry for the delay. I already had a comment written, but forgot to send it.

Thank you very much for building this. I have 2 critique points that stop me from me merging:

1. there are many forms of locomotion, so the name for the hook would need to explain the type of locomotion

2. The hook API seems inconsistent, i'd propose something like this
```ts
useLocomotion(controller?: XRHandedness | XRControllerState, translateOptions?: false | { speed?: number, ... }, rotateOptions?: false | { speed?: number, smooth?: number, ... })
```

Thanks for this PR, I will make sure to pay more close attention so we can get this in ASAP 🙏🏼

taeuscherpferd commented:
@bbohlender 

Thanks for your feedback! I've taken it into account and have pushed up a new version

1. I have changed the name to useControllerLocomotion
2. I have updated the API to better match what you've suggested. I left the controller as only type: `XRHandedness` as hooks cannot be called conditionally, so if they did pass a controller in, I'd have to fetch it using the hook anyway, so I figured I'd avoid the extra overhead of calling it twice, and only allow `XRHandedness`. I can still update it to support both types if you would like

I've also updated the API slightly to allow for a callback to be able to use this in physics-based movement scenarios and updated the Minecraft example to reflect this. Excited to hear your thoughts from round 2 :)👍

bbohlender commented:
Awesome modifications, I really like it. Still have a short list of criticism:

1. for hooks in general using multiple parameters is common e.g. `useX(1,2)` is more common then `useX({ a: 1, b: 2 })` therefore I suggest using parameters for the first level.

2. I am not a fan of returning a ref since some uses cases might only need the callback and dont need the ref. I'd therefore suggest providing another parameter called `target` that can either take a ref or a callback.

Jointly through my criticism I'm implicity suggesting the this API:

 ```ts
 useControllerLocomotion(
  movementController?: XRHandedness
  target?: RefObject<Object3D> | ((velocity: Vector3, deltaTime: number, state: RootState, frame: XRFrame | undefined) => void),
  translationOptions?: {
    speed?: number
  },
  rotationOptions?: {
    numberOfDegreesToSnapBy?: number
    viewControlDeadZone?: number
    disableSnapTurning?: boolean
    enableSmoothTurning?: boolean
    smoothTurningSpeed?: number
  })
 ```
 
 Let me know what you think :)

taeuscherpferd commented:
Thanks again, and excellent points 👍

1. That makes total sense, I only grouped them as an object to be more convenient when it was flat, and there were a lot of optional properties to avoid the call looking something like this `useControllerLocomotion(undefined, undefined, 3, undefined, 45)` Now that the options are better grouped, your suggestion makes sense 👍

2. As far as the ref goes, even with the callback, you still need the ref to be able to handle the rotation of the user in VR. I want to say that I agree that making it optional opens the hook up to be used in more use cases, but I'm struggling to think of them. My best attempt would be if the control switched from controlling the user to controlling a vehicle containing the user, that might be a case where this would be useful, but at that point, using the hook to make that work might be more trouble than it's worth. At any rate, I'll give it a shot and see how it feels

taeuscherpferd commented:
All good to go for the next round 👍. I actually am really liking how it's turning out :)

bbohlender commented on packages/react/xr/src/hooks.ts:
```diff
@@ -102,3 +104,132 @@ export function useSessionModeSupported(mode: XRSessionMode, onError?: (error: a
+  rotationOptions?: ControllerLocomotionRotationOptions,
+  movementController?: XRHandedness,
+) {
+  const defaultSpeed = 2
```
these can be declared outside the function

taeuscherpferd replied:
Wasn't sure how you wanted things to be scoped, so I left them in. Moved them out 👍

bbohlender commented on packages/react/xr/src/hooks.ts:
```diff
@@ -102,3 +104,132 @@ export function useSessionModeSupported(mode: XRSessionMode, onError?: (error: a
+export interface ControllerLocomotionTranslationOptions {
+  speed?: number
+}
+export interface ControllerLocomotionRotationOptions {
```
a imo cleaner options variant would be:

```
export type ControllerLocomotionRotationOptions {
viewControlDeadZone?: number
} & ({type?: "snap", degree?: number } | {.type: "smooth", speed?: number  })
```

then it would be also be great if you could document the properties (e.g. `viewControlDeadZone` I have no clue what it is exactly)

taeuscherpferd replied:
I have them all documented in the `all-hooks.md` file, but if you mean through jsdoc, then I'll add that as well 👍

bbohlender reviewed (commented):
Awesome. this already looks very good. I have a few nitpicks and then its good to go for me :)

taeuscherpferd commented:
All of your suggestions have been implemented. Ready for a final look 👍
